File: lib/config/flags.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/config/flags.h"
#include "lib/config/proto.h"
#include "lib/core/utils.h"
#include "lib/core/logger.h"
#include <google/protobuf/text_format.h>
#include <regex>
#include <fstream>

static FlagGroup* currentFlagGroup;
static std::vector<Flag*> all_flags;
static std::map<const std::string, Flag*> flags_by_name;

static void doHelp();
static void doShowConfig();
static void doDoc();

static FlagGroup helpGroup;
static ActionFlag helpFlag = ActionFlag({"--help"}, "Shows the help.", doHelp);

static ActionFlag showConfigFlag = ActionFlag({"--config", "-C"},
    "Shows the currently set configuration and halts.",
    doShowConfig);

static ActionFlag docFlag = ActionFlag(
    {"--doc"}, "Shows the available configuration options and halts.", doDoc);

FlagGroup::FlagGroup()
{
    currentFlagGroup = this;
}

FlagGroup::FlagGroup(std::initializer_list<FlagGroup*> groups): _groups(groups)
{
    currentFlagGroup = this;
}

void FlagGroup::addFlag(Flag* flag)
{
    _flags.push_back(flag);
}

std::vector<std::string> FlagGroup::parseFlagsWithFilenames(int argc,
    const char* argv[],
    std::function<bool(const std::string&)> callback)
{
    if (_initialised)
        throw std::runtime_error("called parseFlags() twice");

    /* Recursively accumulate a list of all flags. */

    all_flags.clear();
    flags_by_name.clear();
    std::function<void(FlagGroup*)> recurse;
    recurse = [&](FlagGroup* group)
    {
        if (group->_initialised)
            return;

        for (FlagGroup* subgroup : group->_groups)
            recurse(subgroup);

        for (Flag* flag : group->_flags)
        {
            for (const auto& name : flag->names())
            {
                if (flags_by_name.find(name) != flags_by_name.end())
                    error("two flags use the name '{}'", name);
                flags_by_name[name] = flag;
            }

            all_flags.push_back(flag);
        }

        group->_initialised = true;
    };
    recurse(this);
    recurse(&helpGroup);

    /* Now actually parse them. */

    std::vector<std::string> filenames;
    int index = 1;
    while (index < argc)
    {
        std::string thisarg = argv[index];
        std::string thatarg = (index < (argc - 1)) ? argv[index + 1] : "";

        std::string key;
        std::string value;
        bool usesthat = false;

        if (thisarg.size() == 0)
        {
            /* Ignore this argument. */
        }
        else if (thisarg[0] != '-')
        {
            /* This is a filename. Pass it to the callback, and if not consumed
             * queue it. */
            if (!callback(thisarg))
                filenames.push_back(thisarg);
        }
        else
        {
            /* This is a flag. */

            if ((thisarg.size() > 1) && (thisarg[1] == '-'))
            {
                /* Long option. */

                auto equals = thisarg.rfind('=');
                if (equals != std::string::npos)
                {
                    key = thisarg.substr(0, equals);
                    value = thisarg.substr(equals + 1);
                }
                else
                {
                    key = thisarg;
                    value = thatarg;
                    usesthat = true;
                }
            }
            else
            {
                /* Short option. */

                if (thisarg.size() > 2)
                {
                    key = thisarg.substr(0, 2);
                    value = thisarg.substr(2);
                }
                else
                {
                    key = thisarg;
                    value = thatarg;
                    usesthat = true;
                }
            }

            auto flag = flags_by_name.find(key);
            if (flag == flags_by_name.end())
            {
                if (beginsWith(key, "--"))
                {
                    std::string path = key.substr(2);
                    if (key.find('.') != std::string::npos)
                    {
                        globalConfig().set(path, value);
                        index += usesthat;
                    }
                    else
                        globalConfig().applyOption(path);
                }
                else
                    error("unrecognised flag '-{}'; try --help", key);
            }
            else
            {
                flag->second->set(value);
                if (usesthat && flag->second->hasArgument())
                    index++;
            }
        }

        index++;
    }

    globalConfig().validateAndThrow();
    return filenames;
}

void FlagGroup::parseFlags(int argc,
    const char* argv[],
    std::function<bool(const std::string&)> callback)
{
    auto filenames = parseFlagsWithFilenames(argc, argv, callback);
    if (!filenames.empty())
        error(
            "non-option parameter '{}' seen (try --help)", *filenames.begin());
}

void FlagGroup::parseFlagsWithConfigFiles(int argc,
    const char* argv[],
    const std::map<std::string, const ConfigProto*>& configFiles)
{
    parseFlags(argc,
        argv,
        [&](const auto& filename)
        {
            globalConfig().readBaseConfigFile(filename);
            return true;
        });
}

void FlagGroup::checkInitialised() const
{
    if (!_initialised)
        throw std::runtime_error("Attempt to access uninitialised flag");
}

Flag::Flag(const std::vector<std::string>& names, const std::string helptext):
    _group(*currentFlagGroup),
    _names(names),
    _helptext(helptext)
{
    if (!currentFlagGroup)
        error("no flag group defined for {}", *names.begin());
    _group.addFlag(this);
}

void BoolFlag::set(const std::string& value)
{
    if ((value == "true") || (value == "y"))
        _value = true;
    else if ((value == "false") || (value == "n"))
        _value = false;
    else
        error("can't parse '{}'; try 'true' or 'false'", value);
    _callback(_value);
    _isSet = true;
}

const std::string HexIntFlag::defaultValueAsString() const
{
    return fmt::format("0x{:x}", _defaultValue);
}

static void doHelp()
{
    std::cout << "FluxEngine options:\n";
    std::cout
        << "Note: options are processed left to right and order matters!\n";
    for (auto flag : all_flags)
    {
        std::cout << "  ";
        bool firstname = true;
        for (auto name : flag->names())
        {
            if (!firstname)
                std::cout << ", ";
            std::cout << name;
            firstname = false;
        }

        if (flag->hasArgument())
            std::cout << " <default: \"" << flag->defaultValueAsString()
                      << "\">";
        std::cout << ": " << flag->helptext() << std::endl;
    }
    exit(0);
}

static void doShowConfig()
{
    std::string s;
    google::protobuf::TextFormat::PrintToString(globalConfig(), &s);
    std::cout << s << '\n';

    exit(0);
}

static void doDoc()
{
    const auto fields = findAllProtoFields(globalConfig().base());
    for (const auto field : fields)
    {
        const std::string& path = field.first;
        const google::protobuf::FieldDescriptor* f = field.second;

        if (f->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE)
            continue;

        std::string helpText = f->options().GetExtension(help);
        std::cout << fmt::format("{}: {}\n", path, helpText);
    }

    exit(0);
}

File: java/com/cowlark/fluxengine/core/Flags.java
package com.cowlark.fluxengine.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Predicate;

import lombok.Builder;

/**
 * Command-line flags system, ported from lib/config/flags.h.
 */
public class Flags
{
    public static class FlagGroup
    {
        private final List<FlagGroup> parents;
        private final List<Flag> flags = new ArrayList<>();
        private boolean initialised;

        public FlagGroup()
        {
            parents = List.of();
        }

        public FlagGroup(FlagGroup... parents)
        {
            this.parents = List.of(parents);
        }

        public void addFlag(Flag flag)
        {
            flags.add(flag);
        }

        public void parse(String[] argv)
        {
            List<String> filenames = parseWithFilenames(argv, unused -> false);
            if (!filenames.isEmpty())
                throw new FluxEngineException(
                    "non-option parameter '" + filenames.get(0) + "' seen (try --help)");
        }

        public List<String> parseWithFilenames(String[] argv, Predicate<String> callback)
        {
            if (initialised)
                throw new IllegalStateException("called parse() twice");

            /* Recursively accumulate a list of all flags. */
            Map<String, Flag> flagsByName = new HashMap<>();
            recurse(this, flagsByName);

            List<String> filenames = new ArrayList<>();
            int index = 0;
            while (index < argv.length)
            {
                String thisArg = argv[index];
                String thatArg = (index < argv.length - 1) ? argv[index + 1] : "";

                String key;
                String value;
                boolean useThat = false;

                if (thisArg.isEmpty())
                {
                    /* Ignore this argument. */
                }
                else if (thisArg.charAt(0) != '-')
                {
                    /* This is a filename. */
                    if (!callback.test(thisArg))
                        filenames.add(thisArg);
                }
                else
                {
                    if (thisArg.length() > 1 && thisArg.charAt(1) == '-')
                    {
                        /* Long option. */
                        int equals = thisArg.lastIndexOf('=');
                        if (equals >= 0)
                        {
                            key = thisArg.substring(0, equals);
                            value = thisArg.substring(equals + 1);
                        }
                        else
                        {
                            key = thisArg;
                            value = thatArg;
                            useThat = true;
                        }
                    }
                    else
                    {
                        /* Short option. */
                        if (thisArg.length() > 2)
                        {
                            key = thisArg.substring(0, 2);
                            value = thisArg.substring(2);
                        }
                        else
                        {
                            key = thisArg;
                            value = thatArg;
                            useThat = true;
                        }
                    }

                    Flag flag = flagsByName.get(key);
                    if (flag == null)
                        throw new FluxEngineException(
                            "unrecognised flag '" + key + "'; try --help");
                    flag.set(value);
                    if (useThat && flag.hasArgument())
                        index++;
                }

                index++;
            }

            return filenames;
        }

        public void checkInitialised()
        {
            if (!initialised)
                throw new IllegalStateException("Attempt to access uninitialised flag");
        }

        private void recurse(FlagGroup group, Map<String, Flag> flagsByName)
        {
            if (group.initialised)
                return;

            for (FlagGroup parent : group.parents)
                recurse(parent, flagsByName);

            for (Flag flag : group.flags)
            {
                for (String name : flag.names())
                {
                    if (flagsByName.containsKey(name))
                        throw new IllegalStateException(
                            "two flags use the name '" + name + "'");
                    flagsByName.put(name, flag);
                }
            }

            group.initialised = true;
        }
    }

    public abstract static class Flag
    {
        private final FlagGroup group;
        private final List<String> names;
        private final String helptext;

        protected Flag(FlagGroup group, List<String> names, String helptext)
        {
            this.group = group;
            this.names = List.copyOf(names);
            this.helptext = helptext;
            group.addFlag(this);
        }

        public String name()
        {
            return names.get(0);
        }

        public List<String> names()
        {
            return names;
        }

        public String helptext()
        {
            return helptext;
        }

        public abstract boolean hasArgument();

        public abstract String defaultValueAsString();

        public abstract void set(String value);

        protected void checkInitialised()
        {
            group.checkInitialised();
        }
    }

    public static class ActionFlag extends Flag
    {
        private final Runnable voidCallback;
        private final Consumer<String> valueCallback;
        private final boolean hasArgument;

        @Builder(setterPrefix = "set")
        private ActionFlag(FlagGroup group, List<String> names, String helpText,
            Runnable voidCallback, Consumer<String> valueCallback)
        {
            super(group, names, helpText);
            this.voidCallback = voidCallback;
            this.valueCallback = valueCallback;
            hasArgument = valueCallback != null;
        }

        @Override
        public boolean hasArgument()
        {
            return hasArgument;
        }

        @Override
        public String defaultValueAsString()
        {
            return "";
        }

        @Override
        public void set(String value)
        {
            if (hasArgument)
                valueCallback.accept(value);
            else
                voidCallback.run();
        }
    }

    public static class SettableFlag extends Flag
    {
        private boolean value;

        @Builder(setterPrefix = "set")
        private SettableFlag(FlagGroup group, List<String> names, String helpText)
        {
            super(group, names, helpText);
        }

        public boolean get()
        {
            checkInitialised();
            return value;
        }

        @Override
        public boolean hasArgument()
        {
            return false;
        }

        @Override
        public String defaultValueAsString()
        {
            return "false";
        }

        @Override
        public void set(String value)
        {
            this.value = true;
        }
    }

    public abstract static class ValueFlag<T> extends Flag
    {
        private T defaultValue;
        private final Consumer<T> callback;
        protected T value;
        private boolean isSet;

        protected ValueFlag(FlagGroup group, List<String> names, String helptext,
            T defaultValue, Consumer<T> callback)
        {
            super(group, names, helptext);
            this.defaultValue = defaultValue;
            this.value = defaultValue;
            this.callback = callback;
        }

        public T get()
        {
            checkInitialised();
            return value;
        }

        public boolean isSet()
        {
            return isSet;
        }

        public void setDefaultValue(T value)
        {
            defaultValue = value;
            this.value = value;
        }

        protected void setValue(T value)
        {
            this.value = value;
            callback.accept(value);
            isSet = true;
        }
    }

    public static class StringFlag extends ValueFlag<String>
    {
        @Builder(setterPrefix = "set")
        private StringFlag(FlagGroup group, List<String> names, String helpText,
            String defaultValue, Consumer<String> callback)
        {
            super(group, names, helpText,
                defaultValue != null ? defaultValue : "",
                callback != null ? callback : unused -> {});
        }

        @Override
        public boolean hasArgument()
        {
            return true;
        }

        @Override
        public String defaultValueAsString()
        {
            return value;
        }

        @Override
        public void set(String value)
        {
            setValue(value);
        }
    }

    public static class IntFlag extends ValueFlag<Integer>
    {
        @Builder(setterPrefix = "set")
        private IntFlag(FlagGroup group, List<String> names, String helpText,
            int defaultValue, Consumer<Integer> callback)
        {
            super(group, names, helpText, defaultValue,
                callback != null ? callback : unused -> {});
        }

        @Override
        public boolean hasArgument()
        {
            return true;
        }

        @Override
        public String defaultValueAsString()
        {
            return Integer.toString(value);
        }

        @Override
        public void set(String value)
        {
            setValue(Integer.parseInt(value));
        }
    }

    public static class HexIntFlag extends ValueFlag<Integer>
    {
        @Builder(setterPrefix = "set")
        private HexIntFlag(FlagGroup group, List<String> names, String helpText,
            Integer defaultValue)
        {
            super(group, names, helpText,
                defaultValue != null ? defaultValue : 0, unused -> {});
        }

        @Override
        public boolean hasArgument()
        {
            return true;
        }

        @Override
        public String defaultValueAsString()
        {
            return String.format("0x%x", value);
        }

        @Override
        public void set(String value)
        {
            setValue(Integer.parseInt(value));
        }
    }

    public static class DoubleFlag extends ValueFlag<Double>
    {
        @Builder(setterPrefix = "set")
        private DoubleFlag(FlagGroup group, List<String> names, String helpText,
            Double defaultValue, Consumer<Double> callback)
        {
            super(group, names, helpText,
                defaultValue != null ? defaultValue : 1.0,
                callback != null ? callback : unused -> {});
        }

        @Override
        public boolean hasArgument()
        {
            return true;
        }

        @Override
        public String defaultValueAsString()
        {
            return Double.toString(value);
        }

        @Override
        public void set(String value)
        {
            setValue(Double.parseDouble(value));
        }
    }

    public static class BoolFlag extends ValueFlag<Boolean>
    {
        @Builder(setterPrefix = "set")
        private BoolFlag(FlagGroup group, List<String> names, String helpText,
            boolean defaultValue, Consumer<Boolean> callback)
        {
            super(group, names, helpText, defaultValue,
                callback != null ? callback : unused -> {});
        }

        @Override
        public boolean hasArgument()
        {
            return true;
        }

        @Override
        public String defaultValueAsString()
        {
            return value ? "true" : "false";
        }

        @Override
        public void set(String value)
        {
            if (value.equals("true") || value.equals("y"))
                setValue(true);
            else if (value.equals("false") || value.equals("n"))
                setValue(false);
            else
                throw new FluxEngineException(
                    "can't parse '" + value + "'; try 'true' or 'false'");
        }
    }

    private Flags()
    {
    }
}
